core/store/src/trie/mem/flexible_data/extension.rs

```diff
@@ -0,0 +1,31 @@
+use super::FlexibleDataHeader;
+
+/// Flexibly-sized data header for a trie extension path (which is simply
+/// a byte array).
+#[repr(C, packed(1))]
+#[derive(Clone, Copy)]
+pub struct EncodedExtensionHeader {
+    length: u16,
+}
+
+impl FlexibleDataHeader for EncodedExtensionHeader {
+    type InputData = Box<[u8]>;
+    type View<'a> = &'a [u8];
+    fn from_input(extension: &Box<[u8]>) -> EncodedExtensionHeader {
+        EncodedExtensionHeader { length: extension.len() as u16 }
```

Comment:
Nit: panic if `extension.len() > u16::MAX`. Or at least a `debug_assert` and a comment about it, if it’s supposed to be guaranteed by other parts of the system. Seems like safety does not depend on it anyway, which is good :)